_posts/2016-05-01-Using-JWT-Tokens-for-State-Transfer.md

```diff
@@ -0,0 +1,98 @@
+---
+layout: post
+author: andreas_evers
+title: 'Using JWT tokens for state transfer'
+image: /img/JWT/jwt-logo.png
+tags: [JWT,State,Cache]
+category: Microservices
+comments: true
+---
+
+At one of our clients, we have been using JWT tokens quite extensibly.
+We even use it to persist state on the client.
+
+## Why persisting state on the client?
+
+When buiding microservices, we need to build so-called "cloud native" applications.
+One of the key tenents of cloud native application design is keeping your services stateless.
+The benefit of having stateless applications is foremost the ability to respond to events by adding or removing instances without needing to significantly reconfigure or change the application.
+More stateless services can easily be added when load suddenly increases, or if an existing stateless service fails, it can simply be replaced with another.
+Hence, resilience and agility, are easier to achieve with stateless services.
+
+Keeping your services stateless means we need to persist our state somewhere else.
+Since we are tranferring state in the REST architectural style, we can use the client to retain our state.
+For scaling purposes this is a great solution, as the client will only ever have to store its own state, and the server will be relieved of the state of all its clients.
+
+At our client we have chosen to use JWT tokens for this state transfer to the client.
+While JWT is primarily intended for authentication and authorization purposes, the specification allows us to add any data we'd like to verify later on.
+
+## Looking good
+
+Imagine the following scenario:
+
+A list of products is fetched from the products microservice.
+The user isn't allowed to view all products, so only those products the user has access to are returned.
+When the user wants to order a product, he sends an order request to the order microservice with the id of the product he wants to order.
+At that moment the orders microservice needs to know whether or not the user is allowed to order access this product, let alone order it.
+Since the rights to access and order are the same, we'd like to reuse the information returned from the first call to the products microservice.
+This flow is illustrated below.
+
+![JWT state transfer]({{ '/img/JWT/jwt-for-state-tranfer.png' | prepend: site.baseurl }})
+
+We could call the products microservice from the orders microservice and count on caching, but that would still be an extra network hop and the cache could potentially be invalidated by the time the user orders the product.
+Using the JWT approach, state is given to the client (the list of product ids the user is allowed to access), and being passed to the server again the moment an order is placed.
+The signature of the JWT token guarantees us that the state is not tampered with while residing on the client.
```

Comment:
The signature of the JWT token guarantees us that the state is not tampered with, while residing on the client